### DISCUSSION TITLE
Eager loading with nested relationships

### DISCUSSION THREAD

**Discussion by elliotdes** [sqlalchemy/sqlalchemy#11121] [Usage Questions] [answered]
How do I set eager loading for nested relationships in models?

Here is a set of example models:

```python
class League(DeclarativeBase):
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    teams: Mapped[list["Team"]] = relationship()

class Team(DeclarativeBase):
    id: Mapped[int] = mapped_column(primary_key=True)
    league_id: Mapped[int] = mapped_column(ForeignKey("league.id"))
    name: Mapped[str]
    players: Mapped[list["Player"]] = relationship()

class Player(DeclarativeBase):
    id: Mapped[int] = mapped_column(primary_key=True)
    team_id: Mapped[int] = mapped_column(ForeignKey("team.id"))
    name: Mapped[str]
    is_playing: Mapped[bool]
```

How would I eagerly load all players that `is_playing` set to `false`?

I do not want to set `lazy=joined` in the model, as the models are used elsewhere as well.

Here is an example of what I have tried:

```python
stmnt = (
    select(League)
    .options(contains_eager(League.teams))
    .join(Team)
    .join(Player)
    .filter(~Player.is_playing)
)
```

When I then try select players in a team, the players have not been eagerly loaded and all players are loaded, no matter if `is_playing` is set to true or false.

**Comment by zzzeek** [comment, marked as answer]
you would want to chain those options

```
options(contains_eager(League.teams).contains_eager(Team.players))
```


in this case I would not use .join() + contains_eager() if you dont have to.  you can use joinedload()

```
stmnt = (
    select(League)
    .options(
          joinedload(League.teams, innerjoin=True).
         joinedload(Team.players.and_(~Player.is_playing), innerjoin=True)
   )
```

https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#adding-criteria-to-loader-options
